addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/CorsScanRule.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2021 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascanrulesAlpha;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.commons.httpclient.URI;
+import org.apache.commons.lang.RandomStringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.AbstractAppPlugin;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.core.scanner.Category;
+import org.parosproxy.paros.network.HttpHeader;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpRequestHeader;
+import org.parosproxy.paros.network.HttpResponseHeader;
+
+/**
+ * The CORS scan rule identifies Cross-Origin Resource Sharing (CORS) support and overly lenient or
+ * buggy implementations
+ *
+ * @author CravateRouge
+ */
+public class CorsScanRule extends AbstractAppPlugin {
+    private static final Logger LOG = LogManager.getLogger(CorsScanRule.class);
+    private static final String RANDOM_NAME = RandomStringUtils.random(8, true, true);
+    private static final String ACAC = "Access-Control-Allow-Credentials";
+
+    @Override
+    public void scan() {
+        URI uri = getBaseMsg().getRequestHeader().getURI();
+        String authority = uri.getEscapedAuthority();
+        String scheme = uri.getScheme();
+        String handyScheme = scheme + "://";
+
+        // Order of likelihood and severity
+        String[] payloads = {
+            handyScheme + RANDOM_NAME + ".com",
+            "null",
+            handyScheme + RANDOM_NAME + "." + authority,
+            handyScheme + authority + "." + RANDOM_NAME + ".com",
+            // URL encoded backtick used to bypass weak Regex matching only alphanumeric chars to
+            // validate the domain: https://www.corben.io/tricky-CORS/
+            handyScheme + authority + "%60" + RANDOM_NAME + ".com",
+            null,
+            handyScheme + authority
+        };
+
+        boolean secScheme = false;
+        if (scheme == "https") {
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [ES: Comparison of String objects using == or != (ES_COMPARING_STRINGS_WITH_EQ)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#es-comparison-of-string-objects-using-or-es-comparing-strings-with-eq).
This code compares java.lang.String objects for reference equality using the == or != operators. Unless both strings are either constants in a source file, or have been interned using the String.intern() method, the same string value may be represented by two different String objects. Consider using the equals(Object) method instead.